extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -483,40 +517,57 @@ class LifecycleTask extends BackbeatTask {
-                    }
+                // The `Key` names of the versions and delete markers are the same.
+                // Compare the `LastModified` timestamps of the versions and delete markers.
+                const versionLastModified = new Date(versions[vIdx].LastModified);
+                const deleteMarkerLastModified = new Date(deleteMarkers[dmIdx].LastModified);
+                const isVersionLastModifiedNewer = versionLastModified > deleteMarkerLastModified;
+                const isDMLastModifiedNewer = deleteMarkerLastModified > versionLastModified;
+
+                if (isVersionLastModifiedNewer) {
+                    sortedList.push(versions[vIdx++]);
+                } else if (isDMLastModifiedNewer) {
+                    sortedList.push(deleteMarkers[dmIdx++]);
                 } else {
-                    // 4. by VersionId, lower number means newer
-                    if (isVersionVidNewer) {
+                    // If the version and the delete marker have the same last modified date
+                    const nullVersion = (versions[vIdx].VersionId === 'null'
+                    || deleteMarkers[dmIdx].VersionId === 'null');
+                    // and either of them is a null version, we cannot decode the version ID,
+                    // so we push the version object onto the sorted list first.
+                    if (nullVersion) {
                         sortedList.push(versions[vIdx++]);
                     } else {
-                        sortedList.push(deleteMarkers[dmIdx++]);
+                        // and neither of them are null, we decode the version IDs and compare them.
+                        // A lower version ID indicates a more recent version.
+                        // What is the purpose of decoding the version ID?
+                        // The version ID string is a combination of the current time in milliseconds
+                        // and the position of the request in that millisecond.
+                        // This means that two version IDs with the same last modified date can be sorted,
+                        // because the position of the request is also included in the version ID.
+                        const { error: decodeVidError, decodedVersionId } =
+                            this._decodeVID(versions[vIdx].VersionId, log);
+                        if (decodeVidError) {
+                            return { error: decodeVidError, sortedList: null };
+                        }
+
+                        const { error: decodedDMError, decodedVersionId: decodedDMId } =
+                            this._decodeVID(deleteMarkers[dmIdx].VersionId, log);
+                        if (decodedDMError) {
+                            return { error: decodedDMError, sortedList: null };
+                        }
+
+                        const isVersionVidNewer = decodedVersionId < decodedDMId;
+                        if (isVersionVidNewer) {
```

Comment:
these push to `sortedList` and incrementation are repeated in multiple places, while they are actually the critical step of the algorithm.

to make core safer & easier to test, would be better to introduce a `bool compare(version, deleteMarker)` function, which would only perform the comparison (and be easy to test all corner cases), and keep the code simple & readable in the algorithm (handling only the progression and update of the list)